docs/man/output.1
.\" Automatically generated by Pandoc 3.2
.\"
.TH "TRACKER\-OUTPUT" "1" "2024/06" "" "Tracker Output Flag Manual"
.SS NAME
tracker \f[B]\-\-output\f[R] \- Control how and where output is printed
.SS SYNOPSIS
tracker \f[B]\-\-output\f[R] <format[:file,\&...]> |
gotemplate=template[:file,\&...]
| forward:url | webhook:url |
option:{stack\-addresses,exec\-env,exec\-hash[={inode,dev\-inode,digest\-inode}],parse\-arguments,parse\-arguments\-fds,sort\-events}
\&...
.SS DESCRIPTION
The \f[B]\-\-output\f[R] flag allows you to control how and where the
output is printed.
.PP
Format options:
.IP \[bu] 2
\f[B]table[:/path/to/file,\&...]\f[R]: Output events in table format.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]table\-verbose[:/path/to/file,\&...]\f[R]: Output events in table
format with extra fields per event.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]json[:/path/to/file,\&...]\f[R]: Output events in JSON format.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]gotemplate=/path/to/template[:/path/to/file,\&...]\f[R]: Output
events formatted using a given Go template file.
The default path to the file is stdout.
Multiple file paths can be specified, separated by commas.
.IP \[bu] 2
\f[B]none\f[R]: Ignore the stream of events output.
This is usually used with the \f[B]\-\-capture\f[R] flag.
.PP
Fluent Forward options:
.IP \[bu] 2
\f[B]forward:url\f[R]: Send events in JSON format using the Forward
protocol to a Fluent receiver.
Specify the URL of the Fluent receiver.
.PP
Webhook options:
.IP \[bu] 2
\f[B]webhook:url\f[R]: Send events in JSON format to the specified
webhook URL.
.PP
Other options:
.IP \[bu] 2
\f[B]option:{stack\-addresses,exec\-env,exec\-hash,parse\-arguments,sort\-events}\f[R]:
Augment output according to the given options.
The default is none.
Multiple options can be specified, separated by commas.
.RS 2
.IP \[bu] 2
\f[B]stack\-addresses\f[R]: Include stack memory addresses for each
event.
.IP \[bu] 2
\f[B]exec\-env\f[R]: When tracing execve/execveat, show the environment
variables that were used for execution.
.IP \[bu] 2
\f[B]exec\-hash\f[R]: When tracing some file related events, show the
file hash (sha256).
.RS 2
.IP \[bu] 2
Affected events: \f[I]sched_process_exec\f[R],
\f[I]shared_object_loaded\f[R]
.IP \[bu] 2
\f[B]inode\f[R] option recalculates the file hash if the inode\[cq]s
creation time (ctime) differs, which can occur in different namespaces
even for identical inode.
This option is performant, but not recommended and should only be used
if container enrichment can\[cq]t be enabled for digest\-inode, and if
performance is preferred over correctness.
.IP \[bu] 2
\f[B]dev\-inode\f[R] (default) option generally offers better
performance compared to the \f[B]inode\f[R] option, as it bypasses the
need for recalculation by associating the creation time (ctime) with the
device (dev) and inode pair.
It\[cq]s recommended if correctness is preferred over performance
without container enrichment.
.IP \[bu] 2
\f[B]digest\-inode\f[R]\[rq] option is the most efficient, as it keys
the hash to a pair consisting of the container image digest and inode.
This approach, however, necessitates container enrichment.
.RE
.IP \[bu] 2
\f[B]parse\-arguments\f[R]: Do not show raw machine\-readable values for
event arguments.
Instead, parse them into human\-readable strings.
.IP \[bu] 2
\f[B]parse\-arguments\-fds\f[R]: Enable parse\-arguments and enrich file
descriptors (fds) with their file path translation.
This can cause pipeline slowdowns.
.IP \[bu] 2
\f[B]sort\-events\f[R]: Enable sorting events before passing them to the
output.
This may decrease the overall program efficiency.
.RE
.SS EXAMPLES
.IP \[bu] 2
To output events as JSON to stdout, use the following flag:
.RS 2
.IP
.EX
\-\-output json
.EE
.RE
.IP \[bu] 2
To output events as JSON to \f[CR]/my/out\f[R], use the following flag:
.RS 2
.IP
.EX
\-\-output json:/my/out
.EE
.RE
.IP \[bu] 2
To output events as the provided Go template to stdout, use the
following flag:
.RS 2
.IP
.EX
\-\-output gotemplate=/path/to/my.tmpl
.EE
.RE
.IP \[bu] 2
To output events as JSON to both \f[CR]/my/out\f[R] and
\f[CR]/my/out2\f[R], use the following flag:
.RS 2
.IP
.EX
\-\-output json:/my/out1,/my/out2
.EE
.RE
.IP \[bu] 2
To ignore events output, use the following flag:
.RS 2
.IP
.EX
\-\-output none
.EE
.RE
.IP \[bu] 2
To output events as a table with stack addresses, use the following
flag:
.RS 2
.IP
.EX
\-\-output table \-\-output option:stack\-addresses
.EE
.RE
.IP \[bu] 2
To output events via the Forward protocol to \f[CR]127.0.0.1\f[R] on
port \f[CR]24224\f[R] with the tag `tracker' using TCP, use the following
flag:
.RS 2
.IP
.EX
\-\-output forward:tcp://user:pass\[at]127.0.0.1:24224?tag=tracker
.EE
.RE
.IP \[bu] 2
To output events to the webhook endpoint \f[CR]http://webhook:8080\f[R],
use the following flag:
.RS 2
.IP
.EX
\-\-output webhook:http://webhook:8080
.EE
.RE
.IP \[bu] 2
To output events to the webhook endpoint \f[CR]http://webhook:8080\f[R]
with a timeout of 5 seconds, use the following flag:
.RS 2
.IP
.EX
\-\-output webhook:http://webhook:8080?timeout=5s
.EE
.RE


pkg/cmd/flags/output.go
package flags

import (
	"errors"
	"fmt"
	"net/url"
	"os"
	"path/filepath"
	"strings"

	"github.com/khulnasoft/tracker/pkg/config"
	"github.com/khulnasoft/tracker/pkg/errfmt"
)

type PrepareOutputResult struct {
	TrackerConfig  *config.OutputConfig
	PrinterConfigs []config.PrinterConfig
}

func PrepareOutput(outputSlice []string, newBinary bool) (PrepareOutputResult, error) {
	outConfig := PrepareOutputResult{}
	trackerConfig := &config.OutputConfig{}

	// outpath:format
	printerMap := make(map[string]string)

	for _, o := range outputSlice {
		outputParts := strings.SplitN(o, ":", 2)

		if strings.HasPrefix(outputParts[0], "gotemplate=") {
			err := parseFormat(outputParts, printerMap, newBinary)
			if err != nil {
				return outConfig, err
			}
			continue
		}

		switch outputParts[0] {
		case "none":
			if len(outputParts) > 1 {
				if newBinary {
					return outConfig, errors.New("none output does not support path. Run 'man output' for more info")
				}

				return outConfig, errors.New("none output does not support path. Use '--output help' for more info")
			}
			printerMap["stdout"] = "ignore"
		case "table", "table-verbose", "json":
			err := parseFormat(outputParts, printerMap, newBinary)
			if err != nil {
				return outConfig, err
			}
		case "forward":
			err := validateURL(outputParts, "forward", newBinary)
			if err != nil {
				return outConfig, err
			}

			printerMap[outputParts[1]] = "forward"
		case "webhook":
			err := validateURL(outputParts, "webhook", newBinary)
			if err != nil {
				return outConfig, err
			}

			printerMap[outputParts[1]] = "webhook"
		case "option":
			err := parseOption(outputParts, trackerConfig, newBinary)
			if err != nil {
				return outConfig, err
			}
		default:
			if newBinary {
				return outConfig, fmt.Errorf("invalid output flag: %s, run 'man output' for more info", outputParts[0])
			}

			return outConfig, fmt.Errorf("invalid output flag: %s, use '--output help' for more info", outputParts[0])
		}
	}

	// default
	if len(printerMap) == 0 {
		printerMap["stdout"] = "table"
	}

	printerConfigs, err := getPrinterConfigs(printerMap, trackerConfig, newBinary)
	if err != nil {
		return outConfig, err
	}

	outConfig.TrackerConfig = trackerConfig
	outConfig.PrinterConfigs = printerConfigs

	return outConfig, nil
}

func PreparePrinterConfig(printerKind string, outputPath string) (config.PrinterConfig, error) {
	outFile := os.Stdout
	var err error

	if outputPath != "stdout" && outputPath != "" && printerKind != "forward" && printerKind != "webhook" {
		outFile, err = CreateOutputFile(outputPath)
		if err != nil {
			return config.PrinterConfig{}, err
		}
	}

	return config.PrinterConfig{
		Kind:    printerKind,
		OutPath: outputPath,
		OutFile: outFile,
	}, nil
}

// setOption sets the given option in the given config
func setOption(cfg *config.OutputConfig, option string, newBinary bool) error {
	switch option {
	case "stack-addresses":
		cfg.StackAddresses = true
	case "exec-env":
		cfg.ExecEnv = true
	case "parse-arguments":
		cfg.ParseArguments = true
	case "parse-arguments-fds":
		cfg.ParseArgumentsFDs = true
		cfg.ParseArguments = true // no point in parsing file descriptor args only
	case "sort-events":
		cfg.EventsSorting = true
	default:
		if strings.HasPrefix(option, "exec-hash") {
			hashExecParts := strings.Split(option, "=")
			if len(hashExecParts) == 1 {
				if option != "exec-hash" {
					goto invalidOption
				}
				// default
				cfg.CalcHashes = config.CalcHashesDevInode
			} else if len(hashExecParts) == 2 {
				hashExecOpt := hashExecParts[1]
				switch hashExecOpt {
				case "none":
					cfg.CalcHashes = config.CalcHashesNone
				case "inode":
					cfg.CalcHashes = config.CalcHashesInode
				case "dev-inode":
					cfg.CalcHashes = config.CalcHashesDevInode
				case "digest-inode":
					cfg.CalcHashes = config.CalcHashesDigestInode
				default:
					goto invalidOption
				}
			} else {
				goto invalidOption
			}

			return nil
		} else {
			goto invalidOption
		}

	invalidOption:
		if newBinary {
			return errfmt.Errorf("invalid output option: %s, run 'man output' for more info", option)
		}

		return errfmt.Errorf("invalid output option: %s, use '--output help' for more info", option)
	}

	return nil
}

// getPrinterConfigs returns a slice of printer.Configs based on the given printerMap
func getPrinterConfigs(printerMap map[string]string, trackerConfig *config.OutputConfig, newBinary bool) ([]config.PrinterConfig, error) {
	printerConfigs := make([]config.PrinterConfig, 0, len(printerMap))

	for outPath, printerKind := range printerMap {
		if printerKind == "table" {
			if err := setOption(trackerConfig, "parse-arguments", newBinary); err != nil {
				return nil, err
			}
		}
		printerCfg, err := PreparePrinterConfig(printerKind, outPath)
		if err != nil {
			return nil, err
		}
		printerConfigs = append(printerConfigs, printerCfg)
	}

	return printerConfigs, nil
}

// parseFormat parses the given format and sets it in the given printerMap
func parseFormat(outputParts []string, printerMap map[string]string, newBinary bool) error {
	// if not file was passed, we use stdout
	if len(outputParts) == 1 {
		outputParts = append(outputParts, "stdout")
	}

	for _, outPath := range strings.Split(outputParts[1], ",") {
		if outPath == "" {
			if newBinary {
				return errfmt.Errorf("format flag can't be empty, run 'man output' for more info")
			}

			return errfmt.Errorf("format flag can't be empty, use '--output help' for more info")
		}

		if _, ok := printerMap[outPath]; ok {
			if newBinary {
				return errfmt.Errorf("cannot use the same path for multiple outputs: %s, run  'man output' for more info", outPath)
			}

			return errfmt.Errorf("cannot use the same path for multiple outputs: %s, use '--output help' for more info", outPath)
		}
		printerMap[outPath] = outputParts[0]
	}

	return nil
}

// parseOption parses the given option and sets it in the given config
func parseOption(outputParts []string, trackerConfig *config.OutputConfig, newBinary bool) error {
	if len(outputParts) == 1 || outputParts[1] == "" {
		if newBinary {
			return errfmt.Errorf("option flag can't be empty, run 'man output' for more info")
		}

		return errfmt.Errorf("option flag can't be empty, use '--output help' for more info")
	}

	for _, option := range strings.Split(outputParts[1], ",") {
		err := setOption(trackerConfig, option, newBinary)
		if err != nil {
			return err
		}
	}

	return nil
}

// creates *os.File for the given path
func CreateOutputFile(path string) (*os.File, error) {
	fileInfo, err := os.Stat(path)
	if err == nil && fileInfo.IsDir() {
		return nil, errfmt.Errorf("cannot use a path of existing directory %s", path)
	}

	dir := filepath.Dir(path)
	if err := os.MkdirAll(dir, 0755); err != nil {
		return nil, errfmt.Errorf("failed to create directory: %v", err)
	}
	file, err := os.Create(path)
	if err != nil {
		return nil, errfmt.Errorf("failed to create output path: %v", err)
	}

	return file, nil
}

// validateURL validates the given URL
// --output [webhook|forward]:[protocol://user:pass@]host:port[?k=v#f]
func validateURL(outputParts []string, flag string, newBinary bool) error {
	if len(outputParts) == 1 || outputParts[1] == "" {
		if newBinary {
			return errfmt.Errorf("%s flag can't be empty, run 'man output' for more info", flag)
		}

		return errfmt.Errorf("%s flag can't be empty, use '--output help' for more info", flag)
	}
	// Now parse our URL using the standard library and report any errors from basic parsing.
	_, err := url.ParseRequestURI(outputParts[1])

	if err != nil {
		if newBinary {
			return errfmt.Errorf("invalid uri for %s output %q. Run 'man output' for more info", flag, outputParts[1])
		}

		return errfmt.Errorf("invalid uri for %s output %q. Use '--output help' for more info", flag, outputParts[1])
	}

	return nil
}
